### DISCUSSION TITLE
Payjoins

### DISCUSSION THREAD

**Discussion by justinmoon** [fedimint/fedimint#1764] [Fedimint Modules]
TODO

**Comment by justinmoon** [comment]
In payjoin protocols the recipient of a transaction appends an input and outputs to a transaction in order to disrupt the "common input heuristic" and gain more privacy.

Currently our support for the payjoins won't be great because all out UTXOs are k/n P2WSH. So you'd only be able to do payjoins with users who had similar UTXOs. 

But after we have a Taproot on-chain wallet, then our k/n multisig UTXOs would look like any other taproot inputs so long as we sign them in the signature aggregation way (think that's called "key path spend"?).

**Comment by DanGould** [comment]
[Payjoin](https://payjoin.github.io) is one of many ways to append bytes to the blockchain such that sats are transferred at the same time that utxos are merged. This preserves user privacy and, can be used to consolidate utxos in the same transaction. Doing so during times of low fees can save absolute fee expenditure. Payjoins would help to keep the line between coins that are inside or outside of a mint blurred and prevent mints and mint users from being censored.

## Regarding Mixed Inputs

More software is supporting mixed-input spends, so this might not be a huge privacy problem. There is a [proposal to allow it in bip78 too](https://github.com/bitcoin/bips/pull/1244).

Taproot would get rid of the mixed input problem entirely. That might be a better solution.

## Send sats out of a Fedimint via payjoin

Sending payjoin requires signing two PSBTs within a time window of a minute or so. How can a mint deliver two rounds of signatures? Maybe using the lightning gateway would be the easiest way to delegate this signing.

BIP21 unified payment uri includes parameters for sending payjoins. The protocol involves two rounds of signing for a sender: Upon receipt of the uri, the sender constructs a typical transaction funding the bitcoin address supplied in the bip21 as fallback if the payjoin fails. They send it over http and wait for a response. The recipient responds with an augmented a payjoin PSBT that includes their input. Since it has changed, the sender needs to sign this altered PSBT and broadcast it.

## Receiving sats into a Fedimint via payjoin

Receiving payjoin requires hosting a secure endpoint, contributing a utxo to consolidate, and signing one payjoin proposal PSBT of incoming sats.

When and how to authorize fedimint signatures is my biggest open question about this deployment. I'm curious what the constraints are.

**Comment by EthnTuttle** [comment]
https://github.com/fedimint/fedimint/discussions/3264

Discord context:
```
what would it take to do PSBTs from a fedimint? I assume that would be handled withint the wallet module. Might look into this but posting here for asyn discussion.

ethanTuttle — 01/26/2024 10:16 PM
https://docs.rs/crate/psbt/latest
(recommended from a friend)

stachurski — 01/26/2024 10:45 PM
this might be what you're after

https://github.com/fedimint/fedimint/blob/0c6d642/modules/fedimint-wallet-server/src/lib.rs#L1422
January 28, 2024

dpc — 01/28/2024 2:44 AM
I don't get what's the PSBT in Fedimint for.

@dpc
I don't get what's the PSBT in Fedimint for.

ethanTuttle — 01/28/2024 7:00 AM
https://payjoindevkit.org/
[7:01 AM]
this is not what it's in fedimint for right now but payjoin seems like a good (perhaps optional) fit for fedimint

dpc — 01/28/2024 2:17 PM
I don't get the details, but I see a "payjoin" now.
[2:18 PM]
Payjoins are great. But ATM we are barely crawling in the wallet department, so this needs to wait.

stachurski — 01/28/2024 8:36 PM
ahh interesting. there's still an open discussion about what approach we should take to improve the wallet, with the main goal of supporting high volumes of on-chain withdrawals. as dpc mentioned, it might be out of scope in the wallet department for the time being, but I think it's worth an investigation and discussion if introducing payjoin can fit within some of the longer term plans

additional context:
last discussion
improvements doc
January 29, 2024

ethanTuttle — Yesterday at 6:32 AM
It look likes a lot of the discussion and consideration is around pegging out of the federation, I think a payjoin would make more sense on the inbound UTXOs.
[6:38 AM]
Also, am I correct in assuming we want most, if not all of the UTXO selection behavior to be automagically done? or would some UTXO interface be desireable? perhaps grouping UTXOs into usage buckets or categories within the automation.
January 30, 2024

stachurski — Today at 11:35 AM
It look likes a lot of the discussion and consideration is around pegging out of the federation, I think a payjoin would make more sense on the inbound UTXOs
interesting! please correct me if I'm misunderstanding, but I believe that would require the federation using at least one of their UTXOs for part of the deposit transaction, which will reduce the wallet's liquidity. we'd also need to enforce a high feerate for deposits since you could attack a federation with a payjoin that would consume a UTXO with a feerate that won't be confirmed. it also seems like the federation could be held hostage if the transaction feerate needs to be bumped using RBF

Also, am I correct in assuming we want most, if not all of the UTXO selection behavior to be automagically done? or would some UTXO interface be desireable? perhaps grouping UTXOs into usage buckets or categories within the automation.
I think it's best the UTXO selection for the federation is completely automated and simple. if any action is required from either a guardian or user, that's a potential to block on-chain liquidity. it might be worth exploring different designs in a new module, but I think the main priority is making the wallet able to handle high volumes on-chain

curious if there are other thoughts cc: @elsirion @dpc
[11:37 AM]
@ethanTuttle certainly interesting questions, thanks for asking! even if it's too early to introduce something like payjoin I think it's valuable to consider

dpc — Today at 11:48 AM
Non-automated utxo selection would just be an extra attack vector.
```

Per https://github.com/fedimint/fedimint/discussions/3264

I think a payjoin UTXO Gateway (UG) could make sense. This could remove the concern of pinning attacks or general muddling with the core ecash backing UTXO set while also offering fedimint users a more private way to enter into a federation. 

Similar to how a LN gateway can be either CLN or LND (or others), a UTXO gateway could server different purposes at a functional level.
I envision the following as possible functions of a UTXO gateway:
1. ecash backing - the standard ~1:1 ecash backing bitcoin
2. Escrow/operations - not part of the ecash backing UTXOs but accessible to guardians to use for operation costs, etc.
3. payjoin - a gateway offering UTXOs specifically for the purposes of payjoining into the fedimint. This could be offered for both inbound and outbound UTXOs.

**Comment by DanGould** [comment]
I've responded to some of these legitimate concerns regarding payjoin receivers on Discord and would like to add them here for future discussion. Fortunately, I believe they are all being addressed in implementation already. All of the attacks described have automated defenses.

> please correct me if I'm misunderstanding, but I believe that [receiving payjoins for peg-in transactions] would require the federation using at least one of their UTXOs for part of the deposit transaction, which will reduce the wallet's liquidity.

Implementing payjoin peg-ins only creates the *opportunity* to use potentially many of their UTXOs as input, not a requirement. Depending on the liquidity constraints, I recommend a federation only add payjoin input while it already has peg-outs queued. In doing so it can fund the peg-outs using the incoming payjoin sender's UTXO in addition to the settled, internal UTXOs via output substitution. In doing so it has a chance to never be encumbered by the fees of that peg-in UTXO. Funds would flow directly from a peg-in to a peg-out in a single batched transaction. Payjoin therefore effectively gives the federation *more* liquidity. This kind of combined peg-in peg-out [payjoin batching is being developed in Galoy's batching service bria](https://github.com/GaloyMoney/bria/issues/266) to take advantage of the opportunity it presents to more effectively manage UTXOs, scale bitcoin, and save fees.

> we'd also need to enforce a high feerate for deposits since you could attack a federation with a payjoin that would consume a UTXO with a feerate that won't be confirmed. it also seems like the federation could be held hostage if the transaction fee rate needs to be bumped using RBF

True. If a federation contributed input to any incoming transaction without regard to fee rate it could get stuck. Fortunately [fee rate negotiation is part of the payjoin specification](https://github.com/bitcoin/bips/blob/master/bip-0078.mediawiki#user-content-span_idfeeoutputspanFee_output) and implemented in Payjoin Dev Kit. Senders typically pay for both sender and receiver input to incentivize an outcome of better privacy. Fees are negotiated automatically based on peer preferences when the protocol runs and fall back to transactions funded entirely by the sender if negotiation fails.

A federation payjoin receiver would only ever sign transactions that were in its benefit to do so. If a peg-in transaction were proposed with too low a fee rate, then the federation can automatically choose not to contribute any inputs to construct a payjoin. Instead, it would proceed with a peg-in funded only with external UTXOs as it does today by broadcasting the "original psbt" fallback.

In an unlikely emergency situation where a transaction did get stuck even after successful fee negotiation, a beneficial UTXO owner, fedimint or client, could always bump the fee using CPFP.

Payjoin trades non-interactive simplicity for some interactive complexity to gain in the ability to batch transaction intents. Just like fedimint makes that interactivity tradeoff to gain federated custody. If interaction fails, payjoin receivers can always fall back on non-interactive transaction construction. I am aware of no situation where Payjoin puts receivers in a position where they sign transactions they would not otherwise voluntarily sign.

**Comment by DanGould** [comment]
After some discussions in Nashville, it seems like fedimint's current system of using a tweak to receive payjoins might not work with transaction cut-through where peg-in transactions can directly pay peg-outs. Since payjoin PSBT will spend some of the payment that would otherwise go to the mint to fund peg-outs, it needs to keep track of the accounting internally, perhaps by keeping track of each peg-out and crediting funds pegged in the fedimint not strictly to an address(es) produced by a tweak.
